cli/tools/registry/pm.rs

```diff
@@ -0,0 +1,291 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+use std::collections::HashMap;
+use std::path::PathBuf;
+
+use deno_ast::TextChange;
+use deno_config::FmtOptionsConfig;
+use deno_core::anyhow::bail;
+use deno_core::anyhow::Context;
+use deno_core::error::AnyError;
+use deno_core::futures::FutureExt;
+use deno_core::futures::StreamExt;
+use deno_core::serde_json;
+use deno_runtime::deno_fetch::reqwest;
+use deno_semver::jsr::JsrPackageReqReference;
+use deno_semver::npm::NpmPackageReqReference;
+use deno_semver::package::PackageReq;
+use jsonc_parser::ast::ObjectProp;
+use jsonc_parser::ast::Value;
+
+use super::api;
+use crate::args::jsr_api_url;
+use crate::args::AddFlags;
+use crate::args::Flags;
+use crate::factory::CliFactory;
+
+struct SelectedPackage {
+  import_name: String,
+  package_name: String,
+  version: String,
+}
+
+enum PackageAndVersion {
+  NotFound(String),
+  Selected(SelectedPackage),
+}
+
+async fn jsr_find_package_and_select_version(
```

Comment:
Slight nitpick, it's always nice to come into a module and be greeted with the most important public function first before the private ones. That helps the code read more top down rather than having to scroll past these non-important functions to get to the most important public one then bounce back up to these.